cmd/kubeadm/app/discovery/discovery.go

```diff
@@ -18,65 +18,47 @@ package discovery
 
 import (
 	"fmt"
-	"io/ioutil"
-	"net/http"
 
-	"k8s.io/client-go/tools/clientcmd"
 	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
-	kubenode "k8s.io/kubernetes/cmd/kubeadm/app/node"
-	tokenutil "k8s.io/kubernetes/cmd/kubeadm/app/util/token"
+	"k8s.io/kubernetes/cmd/kubeadm/app/discovery/file"
+	"k8s.io/kubernetes/cmd/kubeadm/app/discovery/https"
+	"k8s.io/kubernetes/cmd/kubeadm/app/discovery/token"
+	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
 )
 
-// For identifies and executes the desired discovery mechanism.
-func For(d kubeadmapi.Discovery) (*clientcmdapi.Config, error) {
-	switch {
-	case d.File != nil:
-		return runFileDiscovery(d.File)
-	case d.HTTPS != nil:
-		return runHTTPSDiscovery(d.HTTPS)
-	case d.Token != nil:
-		return runTokenDiscovery(d.Token)
-	default:
-		return nil, fmt.Errorf("couldn't find a valid discovery configuration.")
-	}
-}
-
-// runFileDiscovery executes file-based discovery.
-func runFileDiscovery(fd *kubeadmapi.FileDiscovery) (*clientcmdapi.Config, error) {
-	return clientcmd.LoadFromFile(fd.Path)
-}
+const TokenUser = "tls-bootstrap-token-user"
 
-// runHTTPSDiscovery executes HTTPS-based discovery.
-func runHTTPSDiscovery(hd *kubeadmapi.HTTPSDiscovery) (*clientcmdapi.Config, error) {
-	response, err := http.Get(hd.URL)
-	if err != nil {
-		return nil, err
-	}
-	defer response.Body.Close()
+// For returns a KubeConfig object that can be used for doing the TLS Bootstrap with the right credentials
+// Also, before returning anything, it makes sure it can trust the API Server
+func For(d kubeadmapi.Discovery, authnToken string) (*clientcmdapi.Config, error) {
```

Comment:
We should be careful here.  I don't like seeing the authn token here as it will be to easy to introduce a security issue.  We don't trust the server yet so we shouldn't be passing it a bearer token.  If the server is a bad actor it can then steal the token and own the cluster.  Perhaps put some sort of big comment before "GetValidatedClusterInfoObject" to never pass in the authn token?

Other option would be to have the discovery pass not depend on the authn token and build the kubeconfig explicitly after discovery.